components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -20,9 +23,90 @@
 using clp::ir::EightByteEncodedTextAst;
 using clp::ir::FourByteEncodedTextAst;
 using std::string;
+using std::vector;
 
 namespace clp::ffi {
 namespace {
+/**
+ * Callback to handle JSON exceptions. When we cannot throw a captured JSON exception (i.e., in a
+ * destructor), there must be a way to not crash the program but also let the caller be aware of the
+ * exception. This concept plays the role to allow users to customize a callback handler when
+ * catching an exception.
+ */
+template <typename Func>
+concept JsonExceptionCallbackConcept = std::is_invocable_v<Func, nlohmann::json::exception const&>;
+
+/**
+ * Helper class for iterating over a sub-tree of the schema tree in a depth-first manner.
+ * `KeyValuePairLogEvent::serialize_to_json` uses a stack of instances of this class to perform
+ * depth-first serialization of JSON objects. Each instance of this class holds state from the
+ * perspective of a single node in the schema tree. This includes a JSON map representing itself,
+ * which gets populated while traversing over its children. Once all its children have been visited,
+ * the iterator instance gets popped off the DFS stack, and emplaces the contents of its JSON map
+ * into its parent as a subtree.
+ */
+template <JsonExceptionCallbackConcept JsonExceptionCallback>
+class SchemaTreeDfsIterator {
+public:
+    // Constructor
+    SchemaTreeDfsIterator(
+            SchemaTreeNode const* node,
+            vector<SchemaTreeNode::id_t> children,
+            nlohmann::json::object_t* parent,
+            JsonExceptionCallback json_exception_callback
+    )
+            : m_schema_tree_node{node},
+              m_children{std::move(children)},
+              m_curr_child_it{m_children.cbegin()},
+              m_parent{parent},
+              m_map(nlohmann::json::object()),
+              m_json_exception_callback{json_exception_callback} {}
+
+    // Delete copy/move constructor and assignment
+    SchemaTreeDfsIterator(SchemaTreeDfsIterator const&) = delete;
+    SchemaTreeDfsIterator(SchemaTreeDfsIterator&&) = delete;
+    auto operator=(SchemaTreeDfsIterator const&) -> SchemaTreeDfsIterator& = delete;
+    auto operator=(SchemaTreeDfsIterator&&) -> SchemaTreeDfsIterator& = delete;
+
+    // Destructor
+    ~SchemaTreeDfsIterator() {
+        try {
+            // On exit, if the current node is the root, then move the entire tree to the `parent`
+            // to return. Otherwise, construct a subtree in the parent.
```

Comment:
```suggestion
            // If the current node is the root, then replace the `parent` with this node's JSON
            // object. Otherwise, add this node's JSON object as a child of the parent JSON object.
```